File: core-go/report.go
package openwrightcore

import (
	"crypto/ed25519"
	"crypto/sha256"
	"crypto/x509"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"encoding/pem"
	"fmt"
)

// Report assembly + signing, mirroring src/openwright/report.py:build_report and
// src/openwright/signing.py so a Go-produced report verifies under the Python
// verifier (openwright.verify.verify_report) byte-for-byte.

const reportVersion = "1.0.0"

// BoundaryStatement is copied verbatim from src/openwright/report.py — it is part
// of the signed payload, so a single character drift breaks the signature.
const BoundaryStatement = "This report constitutes EVIDENCE that the listed controls were exercised at " +
	"runtime, attested with a tamper-evident Merkle log and an Ed25519 signature. " +
	"It DOES NOT constitute legal compliance, certification, conformity assessment, " +
	"or an audit opinion. Determinations of compliance are reserved for qualified " +
	"auditors, notified bodies, and counsel. Crosswalks are maintainer-authored, " +
	"rigorously cited, and pending qualified legal/standards review."

// KeyID mirrors signing.py:key_id_for — "ed25519:" + first 32 hex of sha256(rawpub).
func KeyID(publicKeyRaw []byte) string {
	sum := sha256.Sum256(publicKeyRaw)
	return "ed25519:" + hex.EncodeToString(sum[:])[:32]
}

// PublicKeyPEM returns the SubjectPublicKeyInfo PEM for a raw Ed25519 public key,
// matching signing.py:public_key_pem (which Python's verifier parses back).
func PublicKeyPEM(publicKeyRaw []byte) (string, error) {
	der, err := x509.MarshalPKIXPublicKey(ed25519.PublicKey(publicKeyRaw))
	if err != nil {
		return "", err
	}
	block := &pem.Block{Type: "PUBLIC KEY", Bytes: der}
	return string(pem.EncodeToMemory(block)), nil
}

// PublicKeyRawFromPEM extracts the raw 32-byte Ed25519 key from a
// SubjectPublicKeyInfo PEM, mirroring signing.py:public_key_raw_from_pem.
func PublicKeyRawFromPEM(pemStr string) ([]byte, error) {
	block, _ := pem.Decode([]byte(pemStr))
	if block == nil {
		return nil, fmt.Errorf("invalid PEM")
	}
	pub, err := x509.ParsePKIXPublicKey(block.Bytes)
	if err != nil {
		return nil, err
	}
	edpub, ok := pub.(ed25519.PublicKey)
	if !ok {
		return nil, fmt.Errorf("not an Ed25519 public key")
	}
	return []byte(edpub), nil
}

// ReportEventInput is one event to place in the report, alongside its committed
// leaf hash (as stored by the ledger; "sha256:" prefix optional — Python stores
// hex without a prefix in the vectors but the verifier strips a prefix anyway).
type ReportEventInput struct {
	Event    *Event
	LeafHash string // hex (no prefix), as the ledger stores it
}

// ReportInputs are the values build_report needs that are not derivable here.
type ReportInputs struct {
	Origin      string
	Timestamp   string // checkpoint + report timestamp (RFC 3339)
	ReportID    string
	GeneratedAt string
	ScopeDesc   string
	ToolVersion string
	CrosswalkID string
	Events      []ReportEventInput
}

// BuildReport assembles and signs a full report dict, byte-compatible with the
// Python verifier. controls/evidence_gaps are empty (a Go producer emits raw
// evidence; crosswalk evaluation stays in Python). It returns the report as an
// ordered-insensitive map (canonicalization sorts keys).
func BuildReport(priv ed25519.PrivateKey, in ReportInputs) (map[string]interface{}, error) {
	pub := priv.Public().(ed25519.PublicKey)
	pubRaw := []byte(pub)
	keyID := KeyID(pubRaw)

	// Leaves + checkpoint root over the committed leaf hashes.
	leaves := make([][]byte, len(in.Events))
	for i, e := range in.Events {
		lh, err := e.Event.LeafHash()
		if err != nil {
			return nil, fmt.Errorf("event %d leaf hash: %w", i, err)
		}
		// Sanity: the committed leaf hash (if provided) must match recomputation.
		if e.LeafHash != "" {
			want := stripSha256Prefix(e.LeafHash)
			if hex.EncodeToString(lh) != want {
				return nil, fmt.Errorf("event %d committed leaf_hash mismatch", i)
			}
		}
		leaves[i] = lh
	}
	treeSize := len(leaves)
	root := TreeHash(leaves)
	rootHex := hex.EncodeToString(root)

	// Sign the checkpoint.
	cpSig, err := SignCheckpoint(priv, in.Origin, treeSize, rootHex, in.Timestamp)
	if err != nil {
		return nil, err
	}
	checkpoint := map[string]interface{}{
		"origin":        in.Origin,
		"tree_size":     treeSize,
		"root_hash":     rootHex,
		"timestamp":     in.Timestamp,
		"public_key_id": keyID,
		"signature":     base64.StdEncoding.EncodeToString(cpSig),
	}

	// Per-event entries with inclusion proofs pinned to the checkpoint tree_size.
	events := make([]interface{}, 0, len(in.Events))
	var timestamps []string
	seenAgents := map[string]bool{}
	for i, e := range in.Events {
		content, err := e.Event.leafContent()
		if err != nil {
			return nil, err
		}
		proof := InclusionProof(leaves, i)
		entry := map[string]interface{}{
			"event":           content,
			"leaf_index":      i,
			"leaf_hash":       e.LeafHash,
			"inclusion_proof": hexProofList(proof),
		}
		events = append(events, entry)
		timestamps = append(timestamps, e.Event.Timestamp)
		seenAgents[e.Event.Actor.AgentID] = true
	}

	period := map[string]interface{}{}
	if len(timestamps) > 0 {
		min, max := timestamps[0], timestamps[0]
		for _, t := range timestamps[1:] {
			if t < min {
				min = t
			}
			if t > max {
				max = t
			}
		}
		period = map[string]interface{}{"start": min, "end": max}
	}

	agentIDs := make([]interface{}, 0, len(seenAgents))
	for a := range seenAgents {
		agentIDs = append(agentIDs, a)
	}
	// build_report uses sorted(agent_ids or seen_agents); canonical encoding does
	// not sort list elements, so sort here to match Python.
	sortStrings(agentIDs)

	pemStr, err := PublicKeyPEM(pubRaw)
	if err != nil {
		return nil, err
	}

	report := map[string]interface{}{
		"report_version":     reportVersion,
		"report_id":          in.ReportID,
		"generated_at":       in.GeneratedAt,
		"tool":               map[string]interface{}{"name": "openwright", "version": in.ToolVersion},
		"boundary_statement": BoundaryStatement,
		"scope": map[string]interface{}{
			"description": in.ScopeDesc,
			"agent_ids":   agentIDs,
		},
		"period": period,
		"crosswalk": map[string]interface{}{
			"id": in.CrosswalkID,
		},
		"summary": map[string]interface{}{
			"total":                 0,
			"satisfied":             0,
			"not_satisfied":         0,
			"insufficient_evidence": 0,
		},
		"controls":       []interface{}{},
		"evidence_gaps":  []interface{}{},
		"checkpoint":     checkpoint,
		"events":         events,
		"public_key_pem": pemStr,
	}

	// Sign the report over canonical(report without "signature").
	payload, err := CanonicalBytes(report)
	if err != nil {
		return nil, err
	}
	sig := ed25519.Sign(priv, payload)
	report["signature"] = map[string]interface{}{
		"algorithm":     "ed25519",
		"public_key_id": keyID,
		"signature":     base64.StdEncoding.EncodeToString(sig),
	}
	return report, nil
}

// VerifyReportRoot recomputes the Merkle root from a report's events (via the
// event-model serializer) and checks it matches the checkpoint, then verifies
// the checkpoint Ed25519 signature with the report's embedded public key. This
// is the Go side of the bidirectional interop: Go verifying Python evidence.
func VerifyReportRoot(report map[string]interface{}) (bool, string) {
	cp, ok := report["checkpoint"].(map[string]interface{})
	if !ok {
		return false, "no checkpoint"
	}
	rootHex, _ := cp["root_hash"].(string)
	origin, _ := cp["origin"].(string)
	timestamp, _ := cp["timestamp"].(string)
	treeSize := toInt(cp["tree_size"])
	cpSigB64, _ := cp["signature"].(string)

	pemStr, _ := report["public_key_pem"].(string)
	pubRaw, err := PublicKeyRawFromPEM(pemStr)
	if err != nil {
		return false, "bad public_key_pem: " + err.Error()
	}

	// Recompute the root from the events using the event-model serializer.
	rawEvents, ok := report["events"].([]interface{})
	if !ok {
		return false, "no events"
	}
	leavesByIndex := map[int][]byte{}
	for _, re := range rawEvents {
		item, ok := re.(map[string]interface{})
		if !ok {
			return false, "bad event entry"
		}
		idx := toInt(item["leaf_index"])
		evMap, _ := item["event"].(map[string]interface{})
		ev, err := eventFromMap(evMap)
		if err != nil {
			return false, "event decode: " + err.Error()
		}
		lh, err := ev.LeafHash()
		if err != nil {
			return false, "leaf hash: " + err.Error()
		}
		leavesByIndex[idx] = lh
	}
	leaves := make([][]byte, treeSize)
	for i := 0; i < treeSize; i++ {
		lh, ok := leavesByIndex[i]
		if !ok {
			return false, fmt.Sprintf("missing leaf at index %d", i)
		}
		leaves[i] = lh
	}
	recomputed := hex.EncodeToString(TreeHash(leaves))
	if recomputed != rootHex {
		return false, fmt.Sprintf("root mismatch: recomputed=%s checkpoint=%s", recomputed, rootHex)
	}

	// Verify the checkpoint signature.
	cpSig, err := base64.StdEncoding.DecodeString(cpSigB64)
	if err != nil {
		return false, "bad checkpoint signature b64"
	}
	if !VerifyCheckpoint(ed25519.PublicKey(pubRaw), origin, treeSize, rootHex, timestamp, cpSig) {
		return false, "checkpoint signature failed"
	}
	return true, "root recomputed and checkpoint signature verified"
}

// -- small helpers ------------------------------------------------------------

func stripSha256Prefix(s string) string {
	const p = "sha256:"
	if len(s) > len(p) && s[:len(p)] == p {
		return s[len(p):]
	}
	return s
}

func hexProofList(p [][]byte) []interface{} {
	out := make([]interface{}, len(p))
	for i, h := range p {
		out[i] = hex.EncodeToString(h)
	}
	return out
}

func toInt(v interface{}) int {
	switch x := v.(type) {
	case int:
		return x
	case int64:
		return int(x)
	case float64:
		return int(x)
	case json.Number:
		n, _ := x.Int64()
		return int(n)
	}
	return 0
}

func sortStrings(s []interface{}) {
	for i := 1; i < len(s); i++ {
		for j := i; j > 0; j-- {
			a, _ := s[j-1].(string)
			b, _ := s[j].(string)
			if a <= b {
				break
			}
			s[j-1], s[j] = s[j], s[j-1]
		}
	}
}

// eventFromMap re-encodes a generic map into JSON and decodes it through the
// typed Event unmarshaler, so it benefits from the same exclude_none / extra
// handling used during conformance. The map values may carry json.Number from a
// UseNumber-decoded report; json.Marshal emits those as raw numbers.
func eventFromMap(m map[string]interface{}) (*Event, error) {
	raw, err := json.Marshal(m)
	if err != nil {
		return nil, err
	}
	var e Event
	if err := json.Unmarshal(raw, &e); err != nil {
		return nil, err
	}
	return &e, nil
}

File: src/openwright/report.py
"""Report generation: signed JSON, PDF, OSCAL, SARIF (FR-RPT).

The signed JSON report (FR-RPT-02) is the authoritative artifact and is
self-contained for independent verification (FR-VER-05, FR-ATT-04): it carries
the signed checkpoint, every event (hash-only — no raw payloads), and an
inclusion proof per event. The whole report is *also* signed, because the
control results are computed (not Merkle leaves) and must be tamper-evident too.

Every artifact carries the evidence-not-certification boundary (FR-RPT-07).
"""

from __future__ import annotations

import base64
import uuid
from datetime import datetime, timezone
from typing import Any, Dict, List, Optional

from . import __version__
from .canonical import canonical_bytes, to_rfc3339
from .crosswalk import CrosswalkResult
from .ledger import Ledger
from .signing import Checkpoint, KeySource, public_key_pem

REPORT_VERSION = "1.0.0"

BOUNDARY_STATEMENT = (
    "This report constitutes EVIDENCE that the listed controls were exercised at "
    "runtime, attested with a tamper-evident Merkle log and an Ed25519 signature. "
    "It DOES NOT constitute legal compliance, certification, conformity assessment, "
    "or an audit opinion. Determinations of compliance are reserved for qualified "
    "auditors, notified bodies, and counsel. Crosswalks are maintainer-authored, "
    "rigorously cited, and pending qualified legal/standards review."
)


def _signing_payload(report: Dict[str, Any]) -> bytes:
    unsigned = {k: v for k, v in report.items() if k != "signature"}
    return canonical_bytes(unsigned)


def build_report(
    ledger: Ledger,
    crosswalk_result: CrosswalkResult,
    key: KeySource,
    *,
    scope_description: str,
    agent_ids: Optional[List[str]] = None,
    period: Optional[Dict[str, str]] = None,
    checkpoint: Optional[Checkpoint] = None,
    report_id: Optional[str] = None,
    generated_at: Optional[str] = None,
    principal=None,
    authorizer=None,
) -> Dict[str, Any]:
    """Assemble and sign the JSON attestation report."""
    if authorizer is not None:
        from .authz import Capability

        authorizer.require(principal, Capability.GENERATE_REPORT)
    checkpoint = checkpoint or ledger.checkpoint(key)
    tree_size = checkpoint.tree_size

    events: List[Dict[str, Any]] = []
    timestamps: List[str] = []
    seen_agents = set()
    for i in range(tree_size):
        ev = ledger.get_event(i)
        timestamps.append(ev.timestamp)
        seen_agents.add(ev.actor.agent_id)
        events.append(
            {
                "event": ev.model_dump(mode="json", exclude_none=True),
                "leaf_index": i,
                "leaf_hash": ev.ledger.leaf_hash if ev.ledger else None,
                # Pin proofs to the checkpoint's tree_size so they verify against
                # the signed root even if the ledger grew during report build (F6).
                "inclusion_proof": ledger.inclusion_proof_hex(i, tree_size=tree_size),
            }
        )

    if period is None and timestamps:
        period = {"start": min(timestamps), "end": max(timestamps)}

    # Surface self-attested evidence gaps at the top level (U1): they are part of
    # the signed payload, so the completeness story is tamper-evident and visible
    # to a reader without re-deriving it from the event list.
    evidence_gaps: List[Dict[str, Any]] = []
    for item in events:
        ev = item["event"]
        attrs = ev.get("attributes") or {}
        if attrs.get("marker") == "evidence_gap":
            evidence_gaps.append(
                {
                    "event_id": ev.get("event_id"),
                    "leaf_index": item["leaf_index"],
                    "dropped_events": attrs.get("dropped_events"),
                    "window_start": attrs.get("window_start"),
                    "window_end": attrs.get("window_end"),
                }
            )

    cr = crosswalk_result
    report: Dict[str, Any] = {
        "report_version": REPORT_VERSION,
        "report_id": report_id or f"rpt_{uuid.uuid4().hex}",
        "generated_at": generated_at or to_rfc3339(datetime.now(timezone.utc)),
        "tool": {"name": "openwright", "version": __version__},
        "boundary_statement": BOUNDARY_STATEMENT,
        "scope": {
            "description": scope_description,
            "agent_ids": sorted(agent_ids or seen_agents),
        },
        "period": period or {},
        "crosswalk": {
            "id": cr.crosswalk_id,
            "title": cr.crosswalk_title,
            "version": cr.crosswalk_version,
            # Pin the exact crosswalk definition (B9/U2): deep-verify recomputes
            # this hash from the crosswalk it loads and refuses on absence/mismatch,
            # so a verdict can't be silently re-derived under a different crosswalk.
            "content_hash": cr.crosswalk_content_hash,
            "source": cr.source,
            "reviewed_as_of": cr.reviewed_as_of,
            "disclaimer": cr.disclaimer,
        },
        "summary": {
            "total": len(cr.controls),
            "satisfied": cr.satisfied,
            "not_satisfied": cr.not_satisfied,
            "insufficient_evidence": cr.insufficient,
        },
        "controls": [c.model_dump(mode="json", exclude_none=True) for c in cr.controls],
        "evidence_gaps": evidence_gaps,
        "checkpoint": checkpoint.model_dump(mode="json"),
        "events": events,
        "public_key_pem": public_key_pem(key.public_key_raw()).decode("ascii"),
    }

    signature = key.sign(_signing_payload(report))
    report["signature"] = {
        "algorithm": "ed25519",
        "public_key_id": key.key_id(),
        "signature": base64.b64encode(signature).decode("ascii"),
    }
    return report


# -- OSCAL assessment-results (FR-RPT-03) -------------------------------------

# OSCAL finding.target.status.state has only {satisfied, not-satisfied}; it has
# no "insufficient evidence" state. To honor FR-MAP-07 (never collapse the two)
# we map insufficient → not-satisfied with reason "other" AND a `openwright-status`
# prop carrying the exact tri-state, so the distinction is never lost.
_OSCAL_STATE = {
    "satisfied": ("satisfied", "pass"),
    "not_satisfied": ("not-satisfied", "fail"),
    "insufficient_evidence": ("not-satisfied", "other"),
}


def to_oscal(report: Dict[str, Any]) -> Dict[str, Any]:
    now = report["generated_at"]
    findings = []
    observations = []
    for c in report["controls"]:
        obs_uuid = str(uuid.uuid4())
        observations.append(
            {
                "uuid": obs_uuid,
                "description": f"{c['title']}: {c['reason']}",
                "methods": ["TEST"],
                "collected": now,
            }
        )
        state, reason = _OSCAL_STATE[c["status"]]
        findings.append(
            {
                "uuid": str(uuid.uuid4()),
                "title": c["title"],
                "description": c["requirement"],
                "props": [
                    {"name": "openwright-status", "value": c["status"], "ns": "https://github.com/allthingsN/openwright/ns"},
                ],
                "target": {
                    "type": "objective-id",
                    "target-id": c["control_id"],
                    "status": {"state": state, "reason": reason},
                },
                "related-observations": [{"observation-uuid": obs_uuid}],
            }
        )
    return {
        "assessment-results": {
            "uuid": str(uuid.uuid4()),
            "metadata": {
                "title": f"OpenWright evidence — {report['crosswalk']['title']}",
                "last-modified": now,
                "version": report["report_id"],
                "oscal-version": "1.1.3",
                "remarks": BOUNDARY_STATEMENT,
            },
            "import-ap": {"href": "#openwright-evidence"},
            "results": [
                {
                    "uuid": str(uuid.uuid4()),
                    "title": "Runtime control evidence",
                    "description": report["scope"]["description"],
                    "start": report.get("period", {}).get("start", now),
                    "reviewed-controls": {"control-selections": [{"include-all": {}}]},
                    "observations": observations,
                    "risks": [
                        {
                            "uuid": str(uuid.uuid4()),
                            "title": "Control gaps",
                            "description": "Controls not satisfied or with insufficient evidence.",
                            "statement": BOUNDARY_STATEMENT,
                            "status": "open",
                        }
                    ],
                    "findings": findings,
                }
            ],
        }
    }


# -- SARIF gap findings (FR-RPT-04) -------------------------------------------

_SARIF_SCHEMA = "https://docs.oasis-open.org/sarif/sarif/v2.1.0/errata01/os/schemas/sarif-schema-2.1.0.json"
_SARIF_LEVEL = {"not_satisfied": "error", "insufficient_evidence": "warning"}


def to_sarif(report: Dict[str, Any]) -> Dict[str, Any]:
    rules = []
    results = []
    for c in report["controls"]:
        if c["status"] == "satisfied":
            continue
        rules.append(
            {
                "id": c["control_id"],
                "name": c["title"],
                "shortDescription": {"text": c["requirement"][:300]},
                "helpUri": (c.get("citation") or {}).get("url"),
            }
        )
        results.append(
            {
                "ruleId": c["control_id"],
                "level": _SARIF_LEVEL.get(c["status"], "warning"),
                "message": {"text": f"{c['status']}: {c['reason']}"},
                "locations": [
                    {
                        "logicalLocations": [
                            {"name": c["control_id"], "kind": "control"}
                        ]
                    }
                ],
            }
        )
    return {
        "$schema": _SARIF_SCHEMA,
        "version": "2.1.0",
        "runs": [
            {
                "tool": {
                    "driver": {
                        "name": "openwright",
                        "version": __version__,
                        "informationUri": "https://github.com/allthingsN/openwright",
                        # The evidence-not-certification boundary travels with the
                        # SARIF artifact too (FR-RPT-07: every generated report).
                        "fullDescription": {"text": BOUNDARY_STATEMENT},
                        "rules": rules,
                    }
                },
                "results": results,
                "properties": {"openwright_boundary_statement": BOUNDARY_STATEMENT},
            }
        ],
    }


# -- PDF (FR-RPT-01) ----------------------------------------------------------


def render_pdf(report: Dict[str, Any], path: str) -> None:
    """Render a human-readable attestation PDF."""
    from reportlab.lib import colors
    from reportlab.lib.pagesizes import letter
    from reportlab.lib.styles import getSampleStyleSheet, ParagraphStyle
    from reportlab.lib.units import inch
    from reportlab.platypus import (
        SimpleDocTemplate,
        Paragraph,
        Spacer,
        Table,
        TableStyle,
    )

    styles = getSampleStyleSheet()
    small = ParagraphStyle("small", parent=styles["Normal"], fontSize=8, leading=10)
    boundary_style = ParagraphStyle(
        "boundary", parent=styles["Normal"], fontSize=9, leading=12,
        backColor=colors.HexColor("#fff3cd"), borderColor=colors.HexColor("#ffc107"),
        borderWidth=1, borderPadding=6,
    )
    story: List[Any] = []
    story.append(Paragraph("OpenWright — Agent Evidence Attestation", styles["Title"]))
    story.append(Paragraph("Evidence of controls exercised — NOT a compliance certification", styles["Italic"]))
    story.append(Spacer(1, 0.15 * inch))

    cw = report["crosswalk"]
    meta_rows = [
        ["Report ID", report["report_id"]],
        ["Generated", report["generated_at"]],
        ["Tool", f"{report['tool']['name']} {report['tool']['version']}"],
        ["Scope", report["scope"]["description"]],
        ["Agent(s)", ", ".join(report["scope"]["agent_ids"]) or "—"],
        ["Period", f"{report.get('period',{}).get('start','—')} → {report.get('period',{}).get('end','—')}"],
        ["Crosswalk", f"{cw['title']} (v{cw['version']})"],
        ["Source", cw["source"]],
        ["Crosswalk reviewed", cw["reviewed_as_of"]],
    ]
    t = Table(meta_rows, colWidths=[1.6 * inch, 4.9 * inch])
    t.setStyle(TableStyle([
        ("FONTSIZE", (0, 0), (-1, -1), 9),
        ("VALIGN", (0, 0), (-1, -1), "TOP"),
        ("TEXTCOLOR", (0, 0), (0, -1), colors.HexColor("#555555")),
        ("LINEBELOW", (0, 0), (-1, -1), 0.25, colors.HexColor("#dddddd")),
    ]))
    story.append(t)
    story.append(Spacer(1, 0.15 * inch))

    s = report["summary"]
    story.append(Paragraph(
        f"<b>Summary:</b> {s['satisfied']} satisfied · {s['not_satisfied']} not satisfied · "
        f"{s['insufficient_evidence']} insufficient evidence · {s['total']} controls",
        styles["Normal"],
    ))
    story.append(Spacer(1, 0.1 * inch))

    status_color = {
        "satisfied": colors.HexColor("#d4edda"),
        "not_satisfied": colors.HexColor("#f8d7da"),
        "insufficient_evidence": colors.HexColor("#fff3cd"),
    }
    rows = [["Control", "Status", "Evidence / reason"]]
    style_cmds = [
        ("BACKGROUND", (0, 0), (-1, 0), colors.HexColor("#343a40")),
        ("TEXTCOLOR", (0, 0), (-1, 0), colors.white),
        ("FONTSIZE", (0, 0), (-1, -1), 8),
        ("VALIGN", (0, 0), (-1, -1), "TOP"),
        ("GRID", (0, 0), (-1, -1), 0.25, colors.HexColor("#cccccc")),
    ]
    for i, c in enumerate(report["controls"], start=1):
        cite = c.get("citation", {})
        art = f"Art. {cite.get('article')}" if cite.get("article") else cite.get("clause", "")
        rows.append([
            Paragraph(f"<b>{c['control_id']}</b><br/>{art}", small),
            Paragraph(c["status"].replace("_", " "), small),
            Paragraph(f"{c['reason']} (evaluated {c['evaluated_count']}, ok {c['satisfied_count']})", small),
        ])
        style_cmds.append(("BACKGROUND", (1, i), (1, i), status_color.get(c["status"], colors.white)))
    tbl = Table(rows, colWidths=[1.5 * inch, 1.2 * inch, 3.8 * inch], repeatRows=1)
    tbl.setStyle(TableStyle(style_cmds))
    story.append(tbl)
    story.append(Spacer(1, 0.15 * inch))

    cp = report["checkpoint"]
    story.append(Paragraph("<b>Attestation checkpoint (signed tree head)</b>", styles["Normal"]))
    cp_rows = [
        ["Origin", cp["origin"]],
        ["Tree size", str(cp["tree_size"])],
        ["Root hash", cp["root_hash"]],
        ["Timestamp", cp["timestamp"]],
        ["Signing key", cp["public_key_id"]],
    ]
    cpt = Table(cp_rows, colWidths=[1.3 * inch, 5.2 * inch])
    cpt.setStyle(TableStyle([("FONTSIZE", (0, 0), (-1, -1), 7), ("VALIGN", (0, 0), (-1, -1), "TOP")]))
    story.append(cpt)
    story.append(Spacer(1, 0.2 * inch))

    story.append(Paragraph("<b>Boundary statement</b>", styles["Normal"]))
    story.append(Paragraph(report["boundary_statement"], boundary_style))

    SimpleDocTemplate(path, pagesize=letter, title="OpenWright Attestation").build(story)
